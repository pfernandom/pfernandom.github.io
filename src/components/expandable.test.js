import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Expandable from './expandable';

describe('Expandable', () => {
  function fireToggleEvent(element) {
    // eslint-disable-next-line no-param-reassign
    element.open = !element.open;
    return fireEvent(
      element,
      new Event('toggle', {
        bubbles: true,
        cancelable: true,
      }),
    );
  }

  it('renders correctly, expanded', () => {
    const { container, getByTestId } = render(
      <Expandable title={<span>Hello</span>} expanded>
        Content
      </Expandable>,
    );
    expect(getByTestId('expandable-details')).toBeInTheDocument();
    expect(getByTestId('expandable-details').hasAttribute('open')).toBe(true);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly, collapsed', () => {
    const { container, getByTestId } = render(
      <Expandable title={<span>Hello</span>} expanded={false}>
        Content
      </Expandable>,
    );
    const expandableDetails = getByTestId('expandable-details');
    expect(expandableDetails).toBeInTheDocument();
    expect(expandableDetails.hasAttribute('open')).toBe(false);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('it toggles correctly with toggle event', () => {
    const { getByTestId } = render(
      <Expandable title={<span>Hello</span>} expanded>
        Content
      </Expandable>,
    );
    const expandableDetails = getByTestId('expandable-details');
    expect(expandableDetails.hasAttribute('open')).toBe(true);

    const expandableSummary = getByTestId('expandable-summary');
    expect(expandableSummary).toBeInTheDocument();

    fireToggleEvent(expandableDetails);

    expect(expandableDetails.hasAttribute('open')).toBe(false);

    fireToggleEvent(expandableDetails);

    expect(expandableDetails.hasAttribute('open')).toBe(true);
  });

  it('it toggles correctly by updating props', () => {
    const { rerender, getByTestId } = render(
      <Expandable title={<span>Hello</span>} expanded>
        Content
      </Expandable>,
    );
    const expandableDetails = getByTestId('expandable-details');
    expect(expandableDetails.hasAttribute('open')).toBe(true);

    rerender(
      <Expandable title={<span>Hello</span>} expanded={false}>
        Content
      </Expandable>,
    );

    expect(expandableDetails.hasAttribute('open')).toBe(false);

    rerender(
      <Expandable title={<span>Hello</span>} expanded>
        Content
      </Expandable>,
    );

    expect(expandableDetails.hasAttribute('open')).toBe(true);
  });
});
