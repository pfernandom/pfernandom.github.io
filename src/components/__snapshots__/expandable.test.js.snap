// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expandable renders correctly, collapsed 1`] = `
<details
  data-testid="expandable-details"
>
  <summary
    class="expandable__title 
            false
            
            "
    data-testid="expandable-summary"
  >
    <h3
      class="expandable__heading 
            false"
    >
      <span>
        Hello
      </span>
    </h3>
  </summary>
  <div
    class="expandable__content
            false
            no-print"
    hidden=""
  >
    Content
  </div>
</details>
`;

exports[`Expandable renders correctly, expanded 1`] = `
<details
  data-testid="expandable-details"
  open=""
>
  <summary
    class="expandable__title 
            false
            
            "
    data-testid="expandable-summary"
  >
    <h3
      class="expandable__heading 
            false"
    >
      <span>
        Hello
      </span>
    </h3>
  </summary>
  <div
    class="expandable__content
            false
            "
  >
    Content
  </div>
</details>
`;
